src/auth/linking/producer.ts

```diff
@@ -0,0 +1,294 @@
+import aes from "keystore-idb/lib/aes/index.js"
+import rsa from "keystore-idb/lib/rsa/index.js"
+import utils from "keystore-idb/lib/utils.js"
+import { KeyUse, SymmAlg, HashAlg, CharSize } from "keystore-idb/lib/types.js"
+
+import * as did from "../../did/index.js"
+import * as ucan from "../../ucan/index.js"
+import { impl as auth } from "../implementation.js"
+import { EventEmitter } from "../../common/event-emitter.js"
+import { LinkingError, LinkingWarning, handleLinkingError, tryParseMessage } from "../linking.js"
+
+import type { Maybe, Result } from "../../common/index.js"
+import type { EventListener } from "../../common/event-emitter.js"
+
+export type AccountLinkingProvider = {
+  on: OnChallenge & OnLink & OnDone
+  cancel: () => void
+}
+
+type OnChallenge = (
+  event: "challenge",
+  listener: (
+    args: {
+      pin: number[]
+      confirmPin: () => void
+      rejectPin: () => void
+    }
+  ) => void
+) => void
+type OnLink = (event: "link", listener: (args: { approved: boolean; username: string }) => void) => void
+type OnDone = (event: "done", listener: () => void) => void
+
+type LinkingStep = "BROADCAST" | "NEGOTIATION" | "DELEGATION"
+
+type LinkingState = {
+  username: Maybe<string>
+  sessionKey: Maybe<CryptoKey>
+  temporaryRsaPair: Maybe<CryptoKeyPair>
+  step: Maybe<LinkingStep>
+}
+
+/**
+ * Create an account linking provider
+ *
+ * @param options provider options
+ * @param options.username username of the account
+ * @returns an account linking event emitter and cancel function
+ */
+export const createProducer = async (options: { username: string }): Promise<AccountLinkingProvider> => {
+  const { username } = options
+  const canDelegate = await auth.checkCapability(username)
+
+  if (!canDelegate) {
+    throw new LinkingError(`Producer cannot delegate for username ${username}`)
+  }
+
+  let eventEmitter: Maybe<EventEmitter> = new EventEmitter()
+  const ls: LinkingState = {
+    username,
+    sessionKey: null,
+    temporaryRsaPair: null,
+    step: "BROADCAST"
+  }
+
+  const handleMessage = async (event: MessageEvent): Promise<void> => {
+    const { data } = event
+    const message = data.arrayBuffer ? new TextDecoder().decode(await data.arrayBuffer()) : data
+
+    if (ls.step === "BROADCAST") {
```

Comment:
Minor note, might be useful to make these steps (eg. "BROADCAST") constants, so the compiler will detect typos? Also, is there a reason the `Steps` type is redefined in this module, rather than defining it one place. Is it so you expose as little as possible of the internals?

Reply:
> Minor note, might be useful to make these steps (eg. "BROADCAST") constants, so the compiler will detect typos? 

I'm not sure if I understand. Do you mean making them enums? We have the `LinkingStep` type

https://github.com/fission-suite/webnative/blob/a2cb14a2b9c493329bdb207fa5b5cdb524597763/src/auth/linking/consumer.ts#L27

which gives us some checks on what gets assigned to `ls.step`.

> Also, is there a reason the `Steps` type is redefined in this module, rather than defining it one place. Is it so you expose as little as possible of the internals?

No good reason! They turned out the same in the producer and consumer. I'll move them out to `linking.ts` one level up.

Reply:
> which gives us some checks on what gets assigned to `ls.step`

Yeah true, not sure it matters that much regarding type checking.
Enums are indeed one way to approach this. I was thinking of this:
```ts
enum Steps { Broadcast = "BROADCAST",  Negotiation = "NEGOTIATION" }
type LinkingSteps = Broadcast | Negotiation
```

Reply:
Good idea, I've converted this to use an `enum`. The `enum` version reads cleaner and safer than the string version. Thanks for the suggestion. 🙏